docs/GroupDefs.hpp

```diff
@@ -119,7 +119,214 @@
 
 /*!
  * \defgroup DataBoxGroup DataBox
- * \brief Contains (meta)functions used for manipulating DataBoxes
+ * \brief Documentation, functions, metafunctions, and classes necessary for
+ * using DataBox
+ *
+ * DataBox is a heterogeneous compile-time hash map with lazy evaluation of
+ * functions. DataBox can not only store data, but can also store functions that
+ * depend on other data inside the DataBox. The functions will be evaluated when
+ * the data they return is requested. The result is cached, and if a dependency
+ * of the function is modified the cache is invalidated.
+ *
+ * #### Tags and Items
+ *
+ * The compile-time keys are `struct`s called Tags, while the values are called
+ * items. Tags are quite minimal, containing only the information necessary to
+ * store the data and evaluate functions. There are two different types of Tags
+ * that a DataBox can hold: simple tags and compute tags. Simple tags are for
+ * data that is inserted into the DataBox at the time of creation, while compute
+ * tags are for data that will be computed from a function when the compute item
+ * is retrieved. If a compute item is never retrieved from the DataBox then it
+ * is never evaluated.
+ *
+ * Simple tags must have a member type alias `type` that is the type of the data
+ * to be stored, a `static constexpr db::Label label` that is the name of the
+ * tag, and inherit from `db::SimpleTag`. Compute tags must also have a `static
+ * constexpr db::Label label` that is the name of the tag, but they cannot have
+ * a `type` type alias. Instead, compute tags must have a static member function
+ * or static member function pointer named `function`. `function` can be a
+ * function template if necessary. The arguments to the function are retrieved
+ * using tags from the DataBox that the compute tag is in. The tags for the
+ * arguments are set in the member type alias `argument_tags`, which must be a
+ * `tmpl::list` of the tags corresponding to each argument. Note that the order
+ * of the tags in the `argument_list` is the order that they will be passed to
+ * the function. Compute tags must inherit from `db::ComputeTag`.
+ *
+ * Here is an example of a simple tag:
+ *
+ * \snippet Test_DataBox.cpp databox_tag_example
+ *
+ * and an example of a compute tag with a function pointer:
+ *
+ * \snippet Test_DataBox.cpp databox_compute_item_tag_example
+ *
+ * If the compute item's tag is inline then the compute item is of the form:
+ *
+ * \snippet Test_DataBox.cpp compute_item_tag_function
+ *
+ * Finally, a compute tag with an overloaded function and a function template
```

Comment:
Add examples of:
- Overload on the type
- Overload on the number of args
- Simple function template like ComputeLambda0
- Crazy example